### ISSUE TITLE
SAC with gSDE not working on MinitaurBulletEnv-v0

### ISSUE THREAD

**Issue by markub3327** [DLR-RM/rl-baselines3-zoo#117]
Hello @araffin,

  I trained the agent on MinitaurBulletEnv-v0 but without success. I tried to use the Min-max normalization method because every observation has its own range of values, but it's not working and without normalization, I get better results. I test it on my platform [RL Toolkit](https://github.com/markub3327/rl-toolkit), which is based on TF but is almost the same as yours.

Why is MinitaurBulletEnv-v0 not working, when the agent [here](https://www.tensorflow.org/agents/tutorials/7_SAC_minitaur_tutorial) is learning?

My tested hyperparameters:
  | Hyperparameter | Value |
  | ---------------- | ------ |
  | n_timesteps | 1000000 |
  | learning_rate | 0.00073 |
  | batch_size | 256 |
  | buffer_size  | 1000000 |
  | ent_coef | auto |
  | gamma | 0.99 |
  | learning_starts | 10000 |
  | update_interval | 64 |

When you'll find better hyperparameters or anything else please share it with me.

<img width="898" alt="Snímka obrazovky 2021-06-21 o 9 09 42" src="https://user-images.githubusercontent.com/74611856/122723662-82a73400-d273-11eb-9427-20a23991abae.png">

https://user-images.githubusercontent.com/74611856/122724806-9606cf00-d274-11eb-8212-6c414012e29e.mp4

Thanks a lot.

**Comment by araffin** [comment]
Hello,
So after a quick trial, I found those hyperparameters to be working (for unstructured noise):
```yaml
MinitaurBulletEnv-v0:
  n_timesteps: !!float 1e6
  policy: 'MlpPolicy'
  learning_rate: !!float 3e-4
  buffer_size: 100000
  batch_size: 256
  ent_coef: 'auto'
  train_freq: 1
  gradient_steps: 1
  learning_starts: 10000
```

with gSDE (the noise sampling frequency is quite small and therefore quite close to unstructured noise):

```yaml
MinitaurBulletEnv-v0:
  n_timesteps: !!float 1e6
  policy: 'MlpPolicy'
  learning_rate: !!float 3e-4
  buffer_size: 1000000
  batch_size: 256
  ent_coef: 'auto'
  gamma: 0.99
  train_freq: 4
  gradient_steps: 4
  learning_starts: 10000
  use_sde: True
  policy_kwargs: "dict(log_std_init=-3)"
```

As mentioned in the gSDE paper (the arxiv  version will be updated tomorrow: https://arxiv.org/abs/2005.05719 ), the main strength of gSDE is not in simulation but on a real robot (as it reduces wear-and-tear while keeping good performance).

I will try to upload the learning curves and trained agent soon.

**Comment by araffin** [comment]
The training reward with unstructured noise:

![Training_Episodic_Reward](https://user-images.githubusercontent.com/1973948/122769631-02012b80-d2a5-11eb-9a4b-d55d784efb35.png)

**Comment by markub3327** [comment]
I thought that the MinitaurBulletEnv-v0 is Sim-to-Real problematics and gSDE is useful for that too. I will test it with a higher update freq than 64 and a lower learning rate.

Thanks.

**Comment by araffin** [comment]
>I thought that the MinitaurBulletEnv-v0 is Sim-to-Real problematics and gSDE is useful for that too.

gSDE was designed to run RL directly on real robots (no sim2real) even though it should help for sim2real pb too (however, I'm not sure if MinitaurBulletEnv-v0 is Sim-to-Real problematics or not).

**Comment by markub3327** [comment]
I read about Minitaur [here](https://arxiv.org/pdf/1804.10332.pdf). Can you share with me please, Actor loss, Critic loss, Steps (at episode) charts for detailed analysis of my problematics. I try your hyperparameters without significant improvement.

I use client-server architecture for training RL agents (with Reverb).

Thanks.

**Comment by araffin** [comment]
Best for you would be to do a run using SB3 + the rl zoo:
the hyperparams:
```yaml
MinitaurBulletEnv-v0:
  n_timesteps: !!float 1e6
  policy: 'MlpPolicy'
  learning_rate: !!float 3e-4
  buffer_size: 1000000
  batch_size: 256
  ent_coef: 'auto'
  gamma: 0.99
  train_freq: 4
  gradient_steps: 4
  learning_starts: 10000
  use_sde: True
  policy_kwargs: "dict(log_std_init=-3)"
```

Training with tensorboard (logging to `/tmp/tensorboard_sb3/` here)
```
python train.py --algo sac --env MinitaurBulletEnv-v0 -tb /tmp/tensorboard_sb3/ --num-threads 2 --eval-episodes 20 --n-eval-envs 5
```

The learning curves (blue is unstructured noise, orange with gSDE):
Note that the true performance of SAC is higher because I display the training reward here, not the evaluation one using the deterministic controller.
![Training_Episodic_Reward](https://user-images.githubusercontent.com/1973948/122932617-2d038200-d36e-11eb-8510-d18959d41ada.png)

>PS: I think that it is because the learner is faster than the agent. When I try for example AntBulletEnv-v0 everything is good, but the agent plays faster than the learner. In this case, the learner has more data to training from.

you mean that you are learning in parallel of data collection?
Yes this can be definitely a problem.

**Comment by araffin** [comment]
Btw, the new version of the gSDE paper is online https://arxiv.org/abs/2005.05719 ;)

A summary of where it is useful:
![pareto_gsde](https://user-images.githubusercontent.com/1973948/122933003-84a1ed80-d36e-11eb-8370-730808ae6540.png)

**Comment by markub3327** [comment]
Yes, the learner instance is learning in parallel to data collection.

Thanks a lot for your time, I will be trying it with your hyperparameters and try to make the learner slower than the agent is collecting data because those configuration is working on other environments.

**Comment by markub3327** [comment]
Yeah, thanks for the answers I think this issue is solved. I must still be working on RL-Toolkit.

**Comment by araffin** [comment]
>Expectation over time means the average over the time in continuity cost equation?

yes.
I'm using a wrapper to compute it:
```python
class ContinuityCostWrapper(gym.Wrapper):
    """
    Add continuity cost to the reward.
    It assumes that the action space is normalized
    and symmetric (actions in [-1, 1]).
    :param env:
    :param weight_continuity:
    :param verbose:
    :param print_freq: Print every n episodes the mean continuity cost
    """

    def __init__(self, env: gym.Env, weight_continuity: float = 0.0, verbose: int = 0, print_freq: int = 1):
        super(ContinuityCostWrapper, self).__init__(env)
        self.last_action = None
        self.weight_continuity = weight_continuity
        self.verbose = verbose
        self.continuity_hist = []
        self.unnormalized_hist = []
        self.n_episodes = 0
        self.print_freq = print_freq

    def reset(self):
        self.last_action = None
        self.n_episodes += 1
        return self.env.reset()

    def step(self, action):
        obs, reward, done, info = self.env.step(action)
        # Continuity cost
        if self.last_action is not None:
            max_delta = 2.0  # for the action space: high - low = 1 - (-1) = 2
            continuity_cost = np.mean((action - self.last_action) ** 2 / max_delta ** 2)
            unnormalized_cost = np.mean((action - self.last_action) ** 2)
            self.continuity_hist.append(continuity_cost)
            self.unnormalized_hist.append(unnormalized_cost)
            continuity_cost = self.weight_continuity * continuity_cost
            self.last_action = action.copy()
        else:
            continuity_cost = 0.0
            self.last_action = action.copy()

        if done:
            continuity_score = 100 * np.mean(self.continuity_hist)
            if self.verbose > 0 and self.n_episodes % self.print_freq == 0:
                print(f"n_step={len(self.continuity_hist)}")
                print(f"Continuity={continuity_score:.5f} +/ {np.std(self.continuity_hist):.5f}")
                # print(f"Unnormalized continuity={np.mean(self.unnormalized_hist):.5f}\n")
            info["continuity_score"] = continuity_score

        reward -= continuity_cost
        info["continuity_cost"] = continuity_cost
        return obs, reward, done, info
```

**Comment by markub3327** [comment]
@araffin Thanks all is working now. I must use a TQC algorithm instead of classic SAC, because the learner was faster than the agent collecting experiences.

TotalInteractions: 1532116 (the agent steps in the environment)
Train step: 999906 (the learner's training steps)

Now is the agent about 500 000 steps before the learner and learning process is more accurate.

![W B Chart 7  7  2021, 22 37 18](https://user-images.githubusercontent.com/74611856/124825715-4c0c3180-df74-11eb-963e-14420116083c.png)


https://user-images.githubusercontent.com/74611856/124826185-d81e5900-df74-11eb-80a2-3b6fb800bdec.mp4
